File: frontend/src/stores/snippets.ts
import { get, writable } from 'svelte/store'
import { backend, emit } from '../services/backend'
import { notify } from './notifications'
import { activeTabId } from './connections'
import { currentToolWindowKind } from '../services/windowing'
import type { Snippet, SnippetInput } from '../types/snippet'

export const snippets = writable<Snippet[]>([])

/** Loads snippets visible for the given host (0 = all). */
export async function loadSnippets(hostId = 0): Promise<void> {
  try {
    snippets.set((await backend.listSnippets(hostId)) ?? [])
  } catch (e) {
    notify('error', `Could not load snippets: ${e instanceof Error ? e.message : String(e)}`)
  }
}

export async function createSnippet(input: SnippetInput): Promise<void> {
  const snippet = await backend.createSnippet(input)
  snippets.update((list) => [...list, snippet].sort((a, b) => a.name.localeCompare(b.name)))
}

export async function updateSnippet(id: number, input: SnippetInput): Promise<void> {
  const snippet = await backend.updateSnippet(id, input)
  snippets.update((list) => list.map((s) => (s.id === id ? snippet : s)).sort((a, b) => a.name.localeCompare(b.name)))
}

export async function deleteSnippet(id: number): Promise<void> {
  await backend.deleteSnippet(id)
  snippets.update((list) => list.filter((s) => s.id !== id))
}

/** Sends a snippet's command (plus a newline to run it) to the active tab. */
export async function runSnippet(snippet: Snippet): Promise<void> {
  if (currentToolWindowKind === 'snippets') {
    await emit('snippet:run', { command: snippet.command })
    return
  }
  const tabId = get(activeTabId)
  if (!tabId) {
    notify('info', 'Open a session first to run a snippet.')
    return
  }
  try {
    await backend.sendToTab(tabId, snippet.command.endsWith('\n') ? snippet.command : snippet.command + '\n')
  } catch (e) {
    notify('error', `Could not run snippet: ${e instanceof Error ? e.message : String(e)}`)
  }
}

File: internal/app/snippets.go
package app

import (
	"fmt"

	"ssh-first/internal/storage"
)

// ListSnippets returns global snippets plus (when hostID != 0) those scoped to
// that host. Pass hostID 0 to list all.
func (a *App) ListSnippets(hostID int64) ([]storage.Snippet, error) {
	if err := a.requireStore(); err != nil {
		return nil, err
	}
	var scope *int64
	if hostID != 0 {
		scope = &hostID
	}
	return a.store.ListSnippets(scope)
}

// CreateSnippet adds a snippet.
func (a *App) CreateSnippet(input storage.SnippetInput) (storage.Snippet, error) {
	if err := a.requireStore(); err != nil {
		return storage.Snippet{}, err
	}
	if input.Name == "" || input.Command == "" {
		return storage.Snippet{}, fmt.Errorf("name and command are required")
	}
	return a.store.CreateSnippet(input)
}

// UpdateSnippet overwrites an existing snippet.
func (a *App) UpdateSnippet(id int64, input storage.SnippetInput) (storage.Snippet, error) {
	if err := a.requireStore(); err != nil {
		return storage.Snippet{}, err
	}
	if input.Name == "" || input.Command == "" {
		return storage.Snippet{}, fmt.Errorf("name and command are required")
	}
	return a.store.UpdateSnippet(id, input)
}

// DeleteSnippet removes a snippet.
func (a *App) DeleteSnippet(id int64) error {
	if err := a.requireStore(); err != nil {
		return err
	}
	return a.store.DeleteSnippet(id)
}

// SendToTab writes arbitrary text to a terminal tab's stdin — used to run a
// snippet in the active session. A trailing newline (to actually execute the
// command) is the caller's responsibility.
func (a *App) SendToTab(tabID, text string) error {
	return a.WriteTerminalInput(tabID, text)
}
